internal/codespaces/grpc/client.go

```diff
@@ -0,0 +1,112 @@
+// - Start a remote JupyterLab server
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/cli/cli/v2/internal/codespaces/grpc/jupyter"
+	"github.com/cli/cli/v2/pkg/liveshare"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	connectionTimeout = 5 * time.Second
+	requestTimeout    = 30 * time.Second
+)
+
+const (
+	codespacesInternalPort        = 16634
+	codespacesInternalSessionName = "CodespacesInternal"
+)
+
+type Client struct {
+	conn          *grpc.ClientConn
+	token         string
+	listener      net.Listener
+	jupyterClient jupyter.JupyterServerHostClient
+}
+
+type liveshareSession interface {
+	KeepAlive(string)
+	OpenStreamingChannel(context.Context, liveshare.ChannelID) (ssh.Channel, error)
+	StartSharing(context.Context, string, int) (liveshare.ChannelID, error)
+}
+
+// Connects to the gRPC server on the given port
+func Connect(ctx context.Context, session liveshareSession, token string) (*Client, error) {
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", 0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen to local port over tcp: %w", err)
+	}
+
+	// Tunnel the remote gRPC server port to the local port
+	localGrpcServerPort := listener.Addr().(*net.TCPAddr).Port
+	internalTunnelClosed := make(chan error, 1)
+	go func() {
+		fwd := liveshare.NewPortForwarder(session, codespacesInternalSessionName, codespacesInternalPort, true)
+		internalTunnelClosed <- fwd.ForwardToListener(ctx, listener)
+	}()
+
+	// Attempt to connect to the given port
+	conn, err := grpc.Dial(fmt.Sprintf("127.0.0.1:%d", localGrpcServerPort), grpc.WithTimeout(connectionTimeout), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		return nil, err
+	}
+
+	g := &Client{
+		conn:          conn,
+		token:         token,
+		listener:      listener,
+		jupyterClient: jupyter.NewJupyterServerHostClient(conn),
+	}
+
+	return g, nil
+}
+
+// Closes the gRPC connection
+func (g *Client) Close() error {
+	// Closing the local listener effectively closes the gRPC connection
+	if err := g.listener.Close(); err != nil {
+		g.conn.Close() // If we fail to close the listener, explicitly close the gRPC connection and ignore any error
+		return fmt.Errorf("failed to close local tcp port listener: %w", err)
+	}
+
+	return nil
+}
+
+// Appends the authentication token to the gRPC context
+func (g *Client) appendMetadata(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "Authorization", "Bearer "+g.token)
+}
+
+// Starts a remote JupyterLab server to allow the user to connect to the codespace via JupyterLab in their browser
+func (g *Client) StartJupyterServer() (port int, serverUrl string, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
```

Comment:
The method should accept a `ctx context.Context` parameter instead of creating a new background context here.